packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -89,7 +90,10 @@ function patch(n1: VNode, n2: VNode, parent: ParentNode, renderer: RendererAPI)
     }
 
     if (process.env.NODE_ENV !== 'production') {
-        if (!isSameVnode(n1, n2)) {
+        // The only scenario where the n1.sel and n2.sel are different is when the constructor
+        // of a dynamic custom element has changed.
+        // ex: <lwc:component lwc:is={ctor}>
+        if (!isSameVnode(n1, n2) && !areDynamicVCustomElements(n1, n2)) {
```

Comment:
I thought about having the new dynamic component vnodes only go through the `updateDynamicChildren` path but I think that it can also fit into the `updateStaticChildren` path, where this will be invoked.

With the new dynamic components, the value of `vnode.sel` will change when the constructor changes meaning this condition will fail in debug mode. 

However, the main reason behind allowing dynamic components to use the `updateStaticChildren` path is that although the value of `sel` is not the same on update, its position within the child array will not change.

Eventually, when the vnode reaches the `patchCustomElement` routine, [it already checks to see if the `ctors` are equivalent and unmounts/mounts them](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/rendering.ts#L380-L387).

This path is triggered whenever the dynamic components constructor is changed (same path that the current `lwc:dynamic` follows):

```html
<template>
    <lwc:component lwc:is={ctor}></lwc:component>
</template>
```
```javascript
import { LightningElement } from 'lwc'

export default class extends LightningElement {
    ctor;
    async connectedCallback() {
        ctor = await import('x/foo');
    }
}

Reply:
It might be worth noting in a comment that we expect the node to be unmounted/remounted when the `sel` changes. Otherwise it might seem like the attrs from one component will be patched onto a different component, which is not at all what's happening here.

Reply:
@nolanlawson I updated the comment in the latest [commit](https://github.com/salesforce/lwc/pull/3338/commits/47973e00fab660af8a3cbd67a5b74c3dbd234b28#diff-8622743d79e146365dd8c59fb85057a09b0f9577d99f065628abdd95d0d7a754R93-R100), let me know if it's clear enough.

Reply:
@caridy would love to get your thoughts on this part of the diffing algo.

Reply:
@jmsjtu Per existing doc, dynamic components should be considered as dynamic for the purposes of diffing. I found these two references
1. https://github.com/salesforce/lwc/blob/v2.37.3/packages/%40lwc/engine-core/src/framework/api.ts#L572
2. https://github.com/salesforce/lwc/pull/1397#discussion_r300906476